### PR TITLE
feat(sequentially): send events to the server sequentially.

### PR DESCRIPTION
**Pull request by Enuviel**

### CODE REVIEW

File: AndroidSDK/src/com/leanplum/internal/Request.java [hunk 1]
```diff
@@ -411,11 +415,10 @@ protected Void doInBackground(Void... params) {
             int statusCode = op.getResponseCode();
 
             Exception errorException = null;
-            if (statusCode >= 400) {
+            if (statusCode == 200) {
+              deleteSentRequests = true;
```
Comment (benmarten) [inline AndroidSDK/src/com/leanplum/internal/Request.java]: why not immediately delete requests here?

File: AndroidSDK/src/com/leanplum/Leanplum.java [hunk 2]
```diff
@@ -573,7 +573,7 @@ public void noPendingDownloads() {
       });
 
       // Reduce latency by running the rest of the start call in a background thread.
-      Util.executeAsyncTask(new AsyncTask<Void, Void, Void>() {
+      Util.executeAsyncTask(true, new AsyncTask<Void, Void, Void>() {
```
Comment (benmarten) [inline AndroidSDK/src/com/leanplum/Leanplum.java]: what does true mean here? can u add javadoc to Util.executeAsyncTask please? the varargs make it hard to understand.....

File: AndroidSDK/src/com/leanplum/Leanplum.java [hunk 3]
```diff
@@ -662,7 +662,7 @@ private static void startHelper(
     Request req = Request.post(Constants.Methods.START, params);
     req.onApiResponse(new Request.ApiResponseCallback() {
       @Override
-      public void response(List<Map<String, Object>> requests, JSONObject response) {
+      public void response(Map<Integer, Map<String, Object>> requests, JSONObject response) {
```
Comment (benmarten) [inline AndroidSDK/src/com/leanplum/Leanplum.java]: why do we need map now?

File: AndroidSDK/src/com/leanplum/internal/Request.java [hunk 4]
```diff
@@ -461,16 +466,72 @@ public void sendEventually() {
...
 
-  static List<Map<String, Object>> popUnsentRequests() {
+  private static void deleteSentRequests(Set<Integer> keySet) {
+    if (keySet == null || keySet.size() == 0) {
+      return;
+    }
+
+    synchronized (lock) {
+      SharedPreferences.Editor editor = getSharedPreferencesEditor();
+      for (int key : keySet) {
```
Comment (benmarten) [inline AndroidSDK/src/com/leanplum/internal/Request.java]: this looks like a lot of overhead for deleting requests

File: AndroidSDK/src/com/leanplum/internal/Request.java [hunk 5]
```diff
@@ -461,16 +466,72 @@ public void sendEventually() {
...
+  private static void saveUnsentRequest(Map<Integer, Map<String, Object>> requestData) {
+    if (requestData == null) {
+      return;
+    }
+    for (Map.Entry<Integer, Map<String, Object>> mapEntry : requestData.entrySet()) {
+      Map<String, Object> args = mapEntry.getValue();
+      Object retryCountString = args.get("retryCount");
+      int retryCount;
+      if (retryCountString != null) {
+        retryCount = Integer.parseInt(retryCountString.toString()) + 1;
```
Comment (benmarten) [inline AndroidSDK/src/com/leanplum/internal/Request.java]: this parsing does not look really performant to me...

File: AndroidSDK/src/com/leanplum/internal/Request.java [hunk 6]
```diff
@@ -482,19 +543,19 @@ public void sendEventually() {
...
       if (count == 0) {
-        return new ArrayList<>();
+        return new TreeMap<>();
       }
       if (remove) {
         editor.remove(Constants.Defaults.COUNT_KEY);
       }
-
-      for (int i = 0; i < count; i++) {
+      int start = preferences.getInt(Constants.Defaults.START_COUNT_KEY, 0);
```
Comment (benmarten) [inline AndroidSDK/src/com/leanplum/internal/Request.java]: I don't get why we need something else than 0 here?

File: AndroidSDK/src/com/leanplum/internal/Util.java [hunk 7]
```diff
@@ -723,9 +724,14 @@ public static boolean isConnected() {
     return CollectionUtil.uncheckedCast(current);
   }
 
-  public static <T> void executeAsyncTask(AsyncTask<T, ?, ?> task, T... params) {
+  public static <T> void executeAsyncTask(boolean singleThread, AsyncTask<T, ?, ?> task,
```
Comment (benmarten) [inline AndroidSDK/src/com/leanplum/internal/Util.java]: javadoc

Comment (benmarten) [inline AndroidSDK/src/com/leanplum/internal/Util.java]: maybe make convenience methods, that way its harder to not confuse them?
e.g. `executeSingleThreadAsyncTask`

File: AndroidSDK/src/com/leanplum/internal/Request.java [hunk 8]
```diff
@@ -530,51 +591,40 @@ public void sendEventually() {
...
-      int retryCount;
-      if (retryCountString != null) {
-        retryCount = Integer.parseInt(retryCountString.toString()) + 1;
-      } else {
-        retryCount = 1;
-      }
-      args.put("retryCount", Integer.toString(retryCount));
-      saveRequestForLater(args);
+    List<Map<String, Object>> allData = new ArrayList<>();
+    for (Integer key : requestData.keySet()) {
```
Comment (benmarten) [inline AndroidSDK/src/com/leanplum/internal/Request.java]: looks like we have to loop over all requests every time on encode, prob. on decode again? can we avoid the loop somehow?

File: AndroidSDK/src/com/leanplum/internal/Request.java [hunk 9]
```diff
@@ -461,6 +461,61 @@ public void sendEventually() {
...
+
+    Context context = Leanplum.getContext();
+    SharedPreferences preferences = context.getSharedPreferences(
+        LEANPLUM, Context.MODE_PRIVATE);
+    SharedPreferences.Editor editor = preferences.edit();
+    int start = preferences.getInt(Constants.Defaults.START_COUNT_KEY, 0);
+
+    for (int i = 0; i < requestData.size(); i++) {
+      Map<String, Object> args = requestData.get(i);
+      Object retryCountString = args.get("retryCount");
```
Comment (benmarten) [inline AndroidSDK/src/com/leanplum/internal/Request.java]: maybe extract retryCount as constant?

File: AndroidSDK/src/com/leanplum/internal/Request.java [hunk 10]
```diff
@@ -461,6 +461,61 @@ public void sendEventually() {
...
+      int retryCount;
+      if (retryCountString != null) {
```
Comment (benmarten) [inline AndroidSDK/src/com/leanplum/internal/Request.java]: potentially shorter:
`retryCount = (retryCountString != null) ? Integer.parseInt(retryCountString.toString()) + 1 : 1`

File: AndroidSDK/src/com/leanplum/internal/Request.java [hunk 11]
```diff
@@ -484,11 +538,11 @@ public void sendEventually() {
...
       }
-      if (remove) {
-        editor.remove(Constants.Defaults.COUNT_KEY);
+      int start = preferences.getInt(Constants.Defaults.START_COUNT_KEY, 0);
+      if (count - start > MAX_ACTIONS_PER_API_CALL) {
+        count = MAX_ACTIONS_PER_API_CALL;
       }
-
-      for (int i = 0; i < count; i++) {
+      for (int i = start; i < count; i++) {
```
Comment (benmarten) [inline AndroidSDK/src/com/leanplum/internal/Request.java]: maybe leave the empty line above?

File: AndroidSDK/src/com/leanplum/internal/Request.java [hunk 12]
```diff
@@ -502,15 +556,16 @@ public void sendEventually() {
           editor.remove(itemKey);
         }
       }
-      if (remove) {
+      if (remove || count == start) {
+        editor.remove(Constants.Defaults.COUNT_KEY);
```
Comment (benmarten) [inline AndroidSDK/src/com/leanplum/internal/Request.java]: is this save to remove if we have more then MAX_ACTIONS_PER_API_CALL events?

File: AndroidSDK/src/com/leanplum/internal/Util.java [hunk 13]
```diff
@@ -723,9 +724,22 @@ public static boolean isConnected() {
...
+      T... params) {
     if (Build.VERSION.SDK_INT >= 11) {
-      task.executeOnExecutor(asyncExecutor, params);
+      if (singleThread) {
+        task.executeOnExecutor(singleThreadExecutor, params);
+      } else {
+        task.executeOnExecutor(asyncExecutor, params);
+      }
     } else {
       task.execute(params);
```
Comment (benmarten) [inline AndroidSDK/src/com/leanplum/internal/Util.java]: on api level 10 is it single thread? if not we should document that too

Comment (Enuviel) [inline AndroidSDK/src/com/leanplum/internal/Util.java]: it's single thread by default.

File: AndroidSDK/src/com/leanplum/internal/Request.java [hunk 14]
```diff
@@ -502,15 +556,16 @@ public void sendEventually() {
...
+        editor.remove(Constants.Defaults.START_COUNT_KEY);
         try {
```
Comment (milos1290) [inline AndroidSDK/src/com/leanplum/internal/Request.java]: move `editor.apply()/editor.commit()` after `if (remove || count == start)`

**Pull request merged**